zipline/assets/asset_writer.py

```diff
@@ -0,0 +1,779 @@
+        """
+
+        self.conn = db_conn
+        self.fuzzy_char = fuzzy_char
+        self.allow_sid_assignment = allow_sid_assignment
+
+        # This flag controls if the AssetDBWriter is allowed to generate its
+        # own sids. If False, metadata that does not contain a sid will raiset
+        # an exception when building assets.
+        if allow_sid_assignment:
+            self.end_date_to_assign = normalize_date(
+                pd.Timestamp('now', tz='UTC'))
+
+        # Create SQL tables
+        self.init_db(self.conn, constraints)
+
+        for identifier in self._data:
+            # Handle case where full Assets are passed in
+            # For example, in the creation of a DataFrameSource, the source's
+            # 'sid' args may be full Assets
+            if isinstance(identifier, Asset):
+                sid = identifier.sid
+                metadata = identifier.to_dict()
+                metadata['asset_type'] = identifier.__class__.__name__
+                self.write_block(sid, **metadata)
+            else:
+                self.write_block(identifier)
+
+
+class NullAssetDBWriterLegacy(AssetDBWriterLegacy):
+    """
+    An implementation of AssetDBWriterLegacy for use
+    when no data is initially specified.
+    """
+
+    def load_data(self, __):
+        for i in iter(()):
+            yield
+
+
+class AssetDBWriterLegacyFromList(AssetDBWriterLegacy):
+    """
+    Returns a generator yielding entries from sid_list.
+    """
+    def load_data(self, sid_list):
+
+        for i in sid_list:
+            yield i
+
+
+class AssetDBWriterLegacyFromDictionary(AssetDBWriterLegacy):
+    """ An implementation of AssetDBWriter for use
+        with dictionaries.
+
+        Expects a dictionary to be passed to load_data
+        with the following format:
+
+        {id_0: {start_date : ...}, id_1: {start_data: ...}, ...}
+    """
+
+    def load_data(self, dict):
+        """
+        Returns a generator yielding pairs of (identifier, metadata)
+        """
+        for identifier, metadata in dict.items():
+            yield identifier, metadata
+
+
+class AssetDBWriterLegacyFromDataFrame(AssetDBWriterLegacy):
+    """ An implementation of AssetDBWriter for use
+        with pandas DataFrames.
+
+        Expects dataframe to be passed to load_data
+        to have the following structure:
+            * column names must be the metadata fields
+            * index must be the different asset identifiers
+            * array contents should be the metadata value
+    """
+
+    def load_data(self, dataframe):
+        """
+        Returns a generator yielding pairs of (identifier, metadata)
+        """
+        for identifier, row in dataframe.iterrows():
+            yield identifier, row.to_dict()
+
+
+class AssetDBWriterLegacyFromReadable(AssetDBWriterLegacy):
+    """ An implementation of AssetDBWriter for use
+        with objects with a 'read' property.
+
+        The object's read method must return rows
```

Comment:
we should rephrase this to `must return an iterable of mapping objects where the mappings have at least one of 'sid' or 'symbol' along with ...`